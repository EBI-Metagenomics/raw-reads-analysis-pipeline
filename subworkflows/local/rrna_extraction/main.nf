include { INFERNAL_CMSEARCH           } from '../../../modules/ebi-metagenomics/infernal/cmsearch/main'
include { CMSEARCHTBLOUTDEOVERLAP     } from '../../../modules/ebi-metagenomics/cmsearchtbloutdeoverlap/main'
include { EASEL_ESLSFETCH             } from '../../../modules/ebi-metagenomics/easel/eslsfetch/main'
include { EXTRACTCOORDS               } from '../../../modules/ebi-metagenomics/extractcoords/main'
include { COMBINEHMMSEARCHTBL } from '../../../modules/local/combinehmmsearchtbl/main'

workflow RRNA_EXTRACTION {

    take:
    ch_fasta     // channel: [ val(meta), [ fasta ] ]
    rfam         // file: rfam for cmsearch
    claninfo     // file: claninfo for cmsearchtbloutdeoverlap

    main:
    ch_versions = Channel.empty()

    ch_fasta = ch_fasta.map { meta, reads ->
        [meta + ['fasta_read_count': reads.countFasta()], reads]
    }
    // ch_fasta.view{ meta, _reads -> "ch_fasta - [${meta.id}, ${meta.platform}, ${meta.single_end}] - ${meta.fasta_read_count}" }
    // decontaminated_reads = decontaminated_reads.filter { meta, _reads ->
    //     meta.decontam_host_read_count > 0
    // }

    ch_chunked_fasta = ch_fasta
        .flatMap{ meta, fasta ->
            def chunks = fasta.splitFasta(file: true, size: 1.MB)
            chunks.collect{ chunk -> tuple(groupKey(meta, chunks.size()), chunk) }
        }

    INFERNAL_CMSEARCH(
        ch_chunked_fasta,
        rfam
    )
    ch_versions = ch_versions.mix(INFERNAL_CMSEARCH.out.versions.first())

    COMBINEHMMSEARCHTBL(
        INFERNAL_CMSEARCH.out.cmsearch_tbl.groupTuple()
    )

    CMSEARCHTBLOUTDEOVERLAP(
        COMBINEHMMSEARCHTBL.out.concatenated_result,
        claninfo
    )
    ch_versions = ch_versions.mix(CMSEARCHTBLOUTDEOVERLAP.out.versions.first())

    ch_easel = ch_fasta
                .join(CMSEARCHTBLOUTDEOVERLAP.out.cmsearch_tblout_deoverlapped)

    EASEL_ESLSFETCH(ch_easel)
    ch_versions = ch_versions.mix(EASEL_ESLSFETCH.out.versions.first())

    EXTRACTCOORDS(
        EASEL_ESLSFETCH.out.easel_coords,
        EASEL_ESLSFETCH.out.matched_seqs_with_coords
    )
    ch_versions = ch_versions.mix(EXTRACTCOORDS.out.versions.first())

    emit:
    cmsearch_deoverlap_out = CMSEARCHTBLOUTDEOVERLAP.out.cmsearch_tblout_deoverlapped   // channel: [ val(meta), [ deoverlapped ] ]
    easel_out = EASEL_ESLSFETCH.out.easel_coords                                        // channel: [ val(meta), [ fasta ] ]
    ssu_fasta = EXTRACTCOORDS.out.ssu_fasta                                             // channel: [ val(meta), [ fasta ] ]
    lsu_fasta = EXTRACTCOORDS.out.lsu_fasta                                             // channel: [ val(meta), [ fasta ] ]
    concat_ssu_lsu_coords = EXTRACTCOORDS.out.concat_ssu_lsu_coords                     // channel: [ val(meta), [ txt ] ]
    versions = ch_versions                                                              // channel: [ versions.yml ]
}

