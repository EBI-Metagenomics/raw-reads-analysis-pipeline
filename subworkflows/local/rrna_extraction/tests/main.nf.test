nextflow_workflow {

    name "Test Subworkflow RRNA_EXTRACTION"
    script "../main.nf"
    workflow "RRNA_EXTRACTION"

    tag "subworkflows"
    tag "subworkflows_"
    tag "subworkflows/rrna_extraction"

    test("mini sample 1 with some matches should match snapshot") {

        when {
            workflow {
		// file("${projectDir}/download_cache/databases/rfam/rfam_models/ribosomal_models/ribo.cm", checkIfExists: true)
		// file("${projectDir}/download_cache/databases/rfam/rfam_models/ribosomal_models/ribo.claninfo", checkIfExists: true)
                """
                input[0] = Channel.of(
		    [
		        ['id': 'test1', 'single_end': false, 'platform': 'wgs'],
	                file("${projectDir}/tests/fixtures/rrna_extraction/ERR10889147.fasta", checkIfExists: true),
		    ]
		)
		input[1] = Channel.value(
		    file("${projectDir}/tests/fixtures/dbs/rfam_mini/rfam_models/ribosomal_models/ribo.cm", checkIfExists: true)
		)
		input[2] = Channel.value(
		    file("${projectDir}/tests/fixtures/dbs/rfam_mini/rfam_models/ribosomal_models/ribo.claninfo", checkIfExists: true)
		)
                """
            }
        }

        then {
	    def lsu_ssu = workflow.out.concat_ssu_lsu_coords.collect{ meta, fp -> [meta, file(fp).text.readLines().sort()] }

            assert workflow.success
            assert snapshot(
                lsu_ssu,
		workflow.out.versions
	    ).match()
        }
    }

    test("mini sample 2 with some matches should match snapshot") {

        when {
            workflow {
                """
                input[0] = Channel.of(
		    [
		        ['id': 'test2', 'single_end': false, 'platform': 'wgs'],
	                file("${projectDir}/tests/fixtures/rrna_extraction/ERR10889056.fasta", checkIfExists: true),
		    ]
		)
		input[1] = Channel.value(
		    file("${projectDir}/tests/fixtures/dbs/rfam_mini/rfam_models/ribosomal_models/ribo.cm", checkIfExists: true)
		)
		input[2] = Channel.value(
		    file("${projectDir}/tests/fixtures/dbs/rfam_mini/rfam_models/ribosomal_models/ribo.claninfo", checkIfExists: true)
		)
                """
            }
        }

        then {
	    def lsu_ssu = workflow.out.concat_ssu_lsu_coords.collect{ meta, fp -> [meta, file(fp).text.readLines().sort()] }

            assert workflow.success
            assert snapshot(
                lsu_ssu,
		workflow.out.versions
	    ).match()
        }
    }

    test("mini sample 1 with no matches should match snapshot") {

        when {
            workflow {
                """
                input[0] = Channel.of(
		    [
		        ['id': 'test1', 'single_end': false, 'platform': 'wgs'],
	                file("${projectDir}/tests/fixtures/rrna_extraction/ERR10889147_nomatch.fasta", checkIfExists: true),
		    ]
		)
		input[1] = Channel.value(
		    file("${projectDir}/tests/fixtures/dbs/rfam_mini/rfam_models/ribosomal_models/ribo.cm", checkIfExists: true)
		)
		input[2] = Channel.value(
		    file("${projectDir}/tests/fixtures/dbs/rfam_mini/rfam_models/ribosomal_models/ribo.claninfo", checkIfExists: true)
		)
                """
            }
        }

        then {
	    def lsu_ssu = workflow.out.concat_ssu_lsu_coords.collect{ meta, fp -> [meta, file(fp).text.readLines().size()] }

            assert workflow.success
            assert snapshot(
                lsu_ssu,
		workflow.out.versions
	    ).match()
        }
    }

    test("mini sample 2 with no matches should match snapshot") {

        when {
            workflow {
                """
                input[0] = Channel.of(
		    [
		        ['id': 'test2', 'single_end': false, 'platform': 'wgs'],
	                file("${projectDir}/tests/fixtures/rrna_extraction/ERR10889056_nomatch.fasta", checkIfExists: true),
		    ]
		)
		input[1] = Channel.value(
		    file("${projectDir}/tests/fixtures/dbs/rfam_mini/rfam_models/ribosomal_models/ribo.cm", checkIfExists: true)
		)
		input[2] = Channel.value(
		    file("${projectDir}/tests/fixtures/dbs/rfam_mini/rfam_models/ribosomal_models/ribo.claninfo", checkIfExists: true)
		)
                """
            }
        }

        then {
	    def lsu_ssu = workflow.out.concat_ssu_lsu_coords.collect{ meta, fp -> [meta, file(fp).text.readLines().size()] }

            assert workflow.success
            assert snapshot(
                lsu_ssu,
		workflow.out.versions
	    ).match()
        }
    }
}
