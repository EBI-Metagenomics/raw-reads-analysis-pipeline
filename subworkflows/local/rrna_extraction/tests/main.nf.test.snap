{
    "mini sample 2 with some matches should match snapshot": {
        "content": [
            [
                [
                    {
                        "groupSize": 1,
                        "groupTarget": {
                            "id": "test2",
                            "single_end": false,
                            "platform": "wgs",
                            "fasta_read_count": 25
                        }
                    },
                    [
                        "ERR10889056.1033653.1-SSU_rRNA_bacteria/1-125 1 125 ERR10889056.1033653.1",
                        "ERR10889056.1033653.2-SSU_rRNA_bacteria/125-1 125 1 ERR10889056.1033653.2",
                        "ERR10889056.1074258.1-LSU_rRNA_bacteria/1-125 1 125 ERR10889056.1074258.1",
                        "ERR10889056.1129341.2-LSU_rRNA_bacteria/1-125 1 125 ERR10889056.1129341.2",
                        "ERR10889056.113454.1-LSU_rRNA_bacteria/125-1 125 1 ERR10889056.113454.1",
                        "ERR10889056.113454.2-LSU_rRNA_bacteria/20-125 20 125 ERR10889056.113454.2",
                        "ERR10889056.1221297.1-SSU_rRNA_bacteria/1-125 1 125 ERR10889056.1221297.1",
                        "ERR10889056.1284882.1-LSU_rRNA_bacteria/37-125 37 125 ERR10889056.1284882.1",
                        "ERR10889056.1284882.2-LSU_rRNA_bacteria/125-1 125 1 ERR10889056.1284882.2",
                        "ERR10889056.1513927.1-LSU_rRNA_bacteria/125-1 125 1 ERR10889056.1513927.1",
                        "ERR10889056.1513927.2-LSU_rRNA_bacteria/1-125 1 125 ERR10889056.1513927.2"
                    ]
                ]
            ],
            [
                "versions.yml:md5,1cd8a1b942aaba58bd5d6ab9c4f3f9b2",
                "versions.yml:md5,806197e595b66087d0516d78eb48ed73",
                "versions.yml:md5,986b50b86874f29db48120870f7ac4f6",
                "versions.yml:md5,d1294ff919ea50bc3a41ddc5a88ab3de"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-05-29T08:47:41.505679132"
    },
    "mini sample 1 with no matches should match snapshot": {
        "content": [
            [
                [
                    {
                        "groupSize": 1,
                        "groupTarget": {
                            "id": "test1",
                            "single_end": false,
                            "platform": "wgs",
                            "fasta_read_count": 13
                        }
                    },
                    2
                ]
            ],
            [
                "versions.yml:md5,1cd8a1b942aaba58bd5d6ab9c4f3f9b2",
                "versions.yml:md5,806197e595b66087d0516d78eb48ed73",
                "versions.yml:md5,986b50b86874f29db48120870f7ac4f6",
                "versions.yml:md5,d1294ff919ea50bc3a41ddc5a88ab3de"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-05-29T08:49:43.849082108"
    },
    "mini sample 2 with no matches should match snapshot": {
        "content": [
            [
                [
                    {
                        "groupSize": 1,
                        "groupTarget": {
                            "id": "test2",
                            "single_end": false,
                            "platform": "wgs",
                            "fasta_read_count": 14
                        }
                    },
                    0
                ]
            ],
            [
                "versions.yml:md5,1cd8a1b942aaba58bd5d6ab9c4f3f9b2",
                "versions.yml:md5,806197e595b66087d0516d78eb48ed73",
                "versions.yml:md5,986b50b86874f29db48120870f7ac4f6",
                "versions.yml:md5,d1294ff919ea50bc3a41ddc5a88ab3de"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-05-29T08:51:39.539640567"
    },
    "mini sample 1 with some matches should match snapshot": {
        "content": [
            [
                [
                    {
                        "groupSize": 1,
                        "groupTarget": {
                            "id": "test1",
                            "single_end": false,
                            "platform": "wgs",
                            "fasta_read_count": 25
                        }
                    },
                    [
                        "ERR10889147.10145.1-LSU_trypano_mito/124-1 124 1 ERR10889147.10145.1",
                        "ERR10889147.10145.1-LSU_trypano_mito/22-125 22 125 ERR10889147.10145.1",
                        "ERR10889147.10145.2-LSU_rRNA_bacteria/1-125 1 125 ERR10889147.10145.2",
                        "ERR10889147.102662.1-LSU_rRNA_bacteria/1-123 1 123 ERR10889147.102662.1",
                        "ERR10889147.108896.1-LSU_rRNA_bacteria/125-1 125 1 ERR10889147.108896.1",
                        "ERR10889147.108896.2-LSU_rRNA_bacteria/1-125 1 125 ERR10889147.108896.2",
                        "ERR10889147.111147.1-LSU_rRNA_bacteria/1-125 1 125 ERR10889147.111147.1",
                        "ERR10889147.111147.2-LSU_rRNA_bacteria/124-1 124 1 ERR10889147.111147.2",
                        "ERR10889147.112245.2-LSU_rRNA_bacteria/124-1 124 1 ERR10889147.112245.2",
                        "ERR10889147.120419.1-LSU_rRNA_bacteria/125-24 125 24 ERR10889147.120419.1",
                        "ERR10889147.120419.2-LSU_rRNA_bacteria/1-125 1 125 ERR10889147.120419.2",
                        "ERR10889147.121522.1-LSU_rRNA_bacteria/125-1 125 1 ERR10889147.121522.1",
                        "ERR10889147.121522.2-LSU_rRNA_bacteria/1-111 1 111 ERR10889147.121522.2",
                        "ERR10889147.55288.2-LSU_trypano_mito/125-1 125 1 ERR10889147.55288.2"
                    ]
                ]
            ],
            [
                "versions.yml:md5,1cd8a1b942aaba58bd5d6ab9c4f3f9b2",
                "versions.yml:md5,806197e595b66087d0516d78eb48ed73",
                "versions.yml:md5,986b50b86874f29db48120870f7ac4f6",
                "versions.yml:md5,d1294ff919ea50bc3a41ddc5a88ab3de"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-05-29T08:43:40.141287196"
    }
}